Adds experimental use of LUIS batch testing

rozele opened the pull request:
Adds a new interface for NLU providers capable of providing batch testing capabilities. Creates an implementation of the batch testing interface for LUIS.

Fixes #334

rozele committed:
Adds experimental use of LUIS batch testing

Adds a new interface for NLU providers capable of providing batch testing capabilities. Creates an implementation of the batch testing interface for LUIS.

Fixes #334

rozele commented on src/NLU.DevOps.Luis.Shared/ILuisBatchTestClient.cs:
```diff
@@ -0,0 +1,39 @@
+        /// <param name="operationId">Operation ID.</param>
+        /// <param name="cancellationToken">Cancellation token.</param>
+        /// <returns>A task to await the operation status.</returns>
+        Task<OperationResponse<string>> GetEvaluationsStatusAsync(string operationId, CancellationToken cancellationToken);
```
This is incorrect, should be Task<OperationResponse<LuisBatchStatusInfo>>

rozele committed:
Adding a few more unsaved files.

rozele committed:
Adds unit tests for batch testing and retry.

rozele committed:
Adds documentation for enabling LUIS batch testing

in4margaret commented on docs/LuisEndpointConfiguration.md:
```diff
@@ -307,6 +338,18 @@ Optional for `test`. When supplied with `true` value, speech-to-text transcripti
+
+### `luisBatchEndpointExperimental`
+
+(Optional) Experimental setting to cconfigure the hostname used by the batch testing interface when the [`luisUseBatchExperimental`](#luisusebatchexperimental) flag is set to `true`.
```
nit: cconfigure -> configure

rozele replied:
Good catch!

in4margaret commented on docs/LuisEndpointConfiguration.md:
```diff
@@ -307,6 +338,18 @@ Optional for `test`. When supplied with `true` value, speech-to-text transcripti
 
+### `luisUseBatchExperimental`
+
+(Optional) Experimental flag that signals that the LUIS batch evaluation endpoint should be used rather than the prediction endpoint.
```
Would be great to mention that this flag is not supported when speech files are used.

rozele replied:
Awesome. Thanks!

in4margaret commented on src/NLU.DevOps.Core.Tests/EnumerableExtensionsTests.cs:
```diff
@@ -0,0 +1,66 @@
+        public static void CreatesBatches()
+        {
+            Enumerable.Repeat(0, 2).Batch(1).Count().Should().Be(2);
+        }
```
nit: would be great to test a case where the number of elements can't be divided by the batch size without remainder.

rozele replied:
Awesome. Thanks.

rozele replied:
Touché.

in4margaret commented on src/NLU.DevOps.Luis.Tests.Shared/RetryTests.cs:
```diff
@@ -0,0 +1,135 @@
+            var delta = 1;
+            var retryAfter = DateTimeOffset.Now.AddSeconds(10).ToString("r", CultureInfo.InvariantCulture);
+            var delay = Retry.GetRetryAfterDelay(retryAfter, TimeSpan.Zero);
+            delay.TotalSeconds.Should().BeApproximately(10, delta);
```
Why do we use delta=1 here, but in previous test it is value 10e-6.

rozele replied:
I actually should not use `BeApproximately` here, because it's +/- delta. The problem with this test is that the implementation of `GetRetryAfterDelay` uses `DateTimeOffset.Now`, rather than some clock interface that can be easily mocked. So, this test is _technically_ non-deterministic because we don't know how much of a diff will be between our estimated clock time (on L31) and the clock time in the `GetRetryAfterDelay` implementation, though given this is a synchronous operation without much compute work, we can be pretty confident that it will be < O(100ms). Thus I chose 1 second as a possible diff value.

It's a good catch though, as this tests 9 <= delay <= 11 (due to `BeApproximately`) when what we really want is 9 <= delay <= 10.

in4margaret commented on src/NLU.DevOps.Luis.Shared/LuisNLUBatchTestClientBase.cs:
```diff
@@ -0,0 +1,296 @@
+                    var matchIndex = jsonObject.Value<int>("matchIndex");
+
+                    // Remove the generic entity properties
+                    jsonObject.Remove("entityType");
```
Should we modify in place or create a new object with fields that we need to have good understanding of what is required?

rozele replied:
We modify in place because the batch API may support additional properties that we don't care about. For example, if the batch API started allowing users to specify a `role` property on each entity, we should pass that through. The only thing we're trying to do is remove any of the generic utterance format properties (`matchText`, `matchIndex` and `entityType`) and replace them with LUIS batch testing format. It's safe to modify in place because we do a deep clone on L129.

in4margaret reviewed (commented):
Did you mean to create a PR to your fork?

rozele commented:
> Did you mean to create a PR to your fork?

@in4margaret - yes, this won't be ready for merge into the main repo for a while.

rozele committed:
Addresses a few review comments.

- Fixes issue with Retry test.
- Adds test for batch operator where item count modulo batch size is non-zero.
- Adds documentation to note that speech and batch testing are incompatible.
- Fixes typo in documentation.
